src/compiler/crystal/macros/interpreter.cr

```diff
@@ -86,6 +86,44 @@ module Crystal
       @vars[name] = value
     end
 
+    def invoke_proc(proc : ProcLiteral, args : Array(ASTNode)) : ASTNode
+      old_vars, @vars = @vars, {} of String => ASTNode
+
+      begin
+        proc.def.args.each_with_index do |param, i|
+          arg = args[i]
+
+          if restriction = param.restriction
+            macro_type = @program.lookup_macro_type(restriction)
+            unless arg.macro_is_a?(macro_type)
+              arg.raise "expected argument ##{i + 1} to 'ProcLiteral#call' to be #{restriction}, not #{arg.class_desc}"
+            end
+          end
```

Comment:
It would be great to add this optional compile time type checking to normal `macro`s as well. But that's for another day 😅.